File: internal/database/models/purpose.go
package models

import "gorm.io/gorm"

type Purpose struct {
	gorm.Model

	Title       string `gorm:"not null"`
	Description string `gorm:"not null"`
	PromptID    uint   `gorm:"not null;index"`
	Prompt      Prompt

	RetentionPeriod PurposeRetentionPeriod `gorm:"constraint:OnDelete:CASCADE;"`
	Limitations     PurposeLimitations     `gorm:"constraint:OnDelete:CASCADE;"`
	Session         Session                `gorm:"constraint:OnDelete:SET NULL;"`
}

type PurposeRetentionPeriod struct {
	gorm.Model
	PurposeID uint `gorm:"uniqueIndex;not null"`

	AudioHours         uint `gorm:"not null;default:0"`
	TranscriptionHours uint `gorm:"not null;default:0"`
	PromptsHours       uint `gorm:"not null;default:0"`
}

type PurposeLimitations struct {
	gorm.Model
	PurposeID uint `gorm:"uniqueIndex;not null"`

	InviteParticipants bool `json:"InviteParticipants"`
	AllowAppRecording  bool `json:"AllowAppRecording"`
}

File: frontend/src/lib/purpose.ts
export type PurposeLimitations = {
  inviteParticipants: boolean;
  allowAppRecording: boolean;
};

export type PurposePrompt = {
  id: number;
  title: string;
};

export type PurposeSummary = {
  id: number;
  title: string;
  description: string;
  promptId: number;
  promptTitle?: string;
  limitations?: PurposeLimitations;
};

export type PurposeApi = {
  id: number;
  title: string;
  description: string;
  prompt_id: number;
  prompt?: PurposePrompt;
  limitations?: {
    invite_participants: boolean;
    allow_app_recording: boolean;
  };
};

export function mapPurpose(api: PurposeApi): PurposeSummary {
  return {
    id: api.id,
    title: api.title,
    description: api.description,
    promptId: api.prompt_id,
    promptTitle: api.prompt?.title,
    limitations: api.limitations
      ? {
          inviteParticipants: api.limitations.invite_participants,
          allowAppRecording: api.limitations.allow_app_recording,
        }
      : undefined,
  };
}

export function isInviteAllowed(purpose?: PurposeSummary | null): boolean {
  return purpose?.limitations?.inviteParticipants ?? true;
}

export function isAppRecordingAllowed(purpose?: PurposeSummary | null): boolean {
  return purpose?.limitations?.allowAppRecording ?? true;
}
